### DISCUSSION TITLE
Help: how to set static keyboard color via WMI/CIM

### DISCUSSION THREAD

**Discussion by TheBestPessimist** [seerge/g-helper#344] [Q&A] [answered]
Hello,  
From your support here and looking at the code, i see that there should be a way to set the keyboard backlight color to my desired RGB.

Something like
```
(Get-WmiObject -Namespace Root/WMI -Class AsusAtkWmi_WMNB).DEVS(0x001234567, MODE_STATIC, R, G, B)
```
Afaiu, there are multiple devices to whom i need to send the correct command?
Would you please be so kind and tell me how that works? 

Rationale: 

I have some issues where the quick keyboard light change breaks me out of flow (for example when working). Even the slow mode of rainbow is extremely fast for me. `¯\_(ツ)_/¯`

I have already implemented a _suuuuper slow rainbow_  using aura creator (it took a while, 😅) [here](https://discourse.tbp.land/t/asus-aura-creator-super-slow-rainbow/4706). It lasts 10 min as that's the maximum length of a light in aura. The problem is that Armoury Crate is stupid and when i turn on my laptop it does not autodetect my keyboard/aura creator, so i need to manually go and enable it in armoury crate. Every. Single. Time. 🙄

As you can imagine that's annoying. So instead, i would like to implement a script (maybe using Autohotkey as i'm using that for so much automation already). 
And the first step is to understand how exactly to set a color from the CLI.
Then i could play around with the time each color is displayed, so that i'm happy with how long the "transitions" take.

If you need more info, my laptop is `Asus ROG Flow X16 GV601VI 2023 RTX 4070`.

Thank you in advance.

**Comment by seerge** [comment, marked as answer]
@TheBestPessimist keyboard is controlled via USB, not via WMI

https://github.com/seerge/g-helper/blob/07020c3561ce26a84ea8130e2720afb4ffb27bb9/app/Aura.cs#L175

Rainbow / any other animation effect has a speed setting (that you can control via app too, under Keyboard -> Extra)

**Comment by TheBestPessimist** [comment]
i dont want to use an animation from Asus though, i want to set static colors myself.
basically this pseudocode in a loop:
- set color 255 0 0
- sleep 10
- set color 255 10 0
- sleep 10
- set color 255 20 0
- sleep 10
- set color 255 30 0
etc.

> keyboard is controlled via USB, not via WMI

that will make things more difficult.

**Comment by TheBestPessimist** [comment]
Is there a difference between `Static` and `Aura Sync Static`? As far as i can see, they have different colors set.

![image](https://user-images.githubusercontent.com/4482210/237062662-5adc4eff-e4c7-428b-a9a2-5461f6313b10.png)


![image](https://user-images.githubusercontent.com/4482210/237063356-66dd5f4d-f8cf-4b69-8f77-736545ed5128.png)

**Comment by seerge** [comment]
@TheBestPessimist i think they are same? Aura sync - is "software" controlled animations (i.e. armoury runs in background and constantly sets new colors to keyboard imitating animations), while basic effects are hardcoded into keyboard itself.

G-helper deliberatedly controls only basic effects.

**Comment by seerge** [comment]
Migrated this to discussions as this is not an issue with the app :)

**Comment by TheBestPessimist** [comment]
Hey, i managed to do something which works for my keyboard: https://github.com/TheBestPessimist/Asus-Keyboard-Super-Slow-Rainbow
It's not slow yet, but it's changing my keyboard's colors, so it's a start. 
_I'm very unfamiliar with C#/.NET etiquette (i'm a kotlin dev) so i probably broke all the style rules._

I'd like to get more of your knowledge though:

When i'm enumerating only my device ID, i still get a list of devices, and out of those only 1 is actually the correct one, even though they all have the same vendor and product.
```
HidDevices.Enumerate(0x0b05, 0x19b6)
```
Is there a way to detect which is the correct device? I find it wasteful to try to set the colors for all the devices which have nothing to do with my keyboard.
What if i want to add support for another keyboard?

**Reply by seerge** [reply]
@TheBestPessimist it could be more than one device on some laptops (like on Z13 it seems like keyboard / logo backlight are separate, but behave in same way). So to be safe i just send payload to all matching devices

**Reply by TheBestPessimist** [reply]
I see. Well, in any case, i added a cache/check for which devices respond to aura commands. That happens on startup, so it should be ok

**Comment by TheBestPessimist** [comment]
Thanks very much for your support. I implemented this!

**Comment by TheBestPessimist** [comment]
@seerge i have one more question: how/where do you detect if the screen is off?

Right now i have the problem that the screen never turns off (ie from idle time) because of sending light commands to keyboard.

_linked issue for future me: https://github.com/TheBestPessimist/Flow.Launcher.Plugin.Asus-Keyboard-Super-Slow-Rainbow/issues/1_

**Reply by seerge** [reply]
Listen to monitor power events https://github.com/seerge/g-helper/blob/f395c706f6a794be9af17fbaec764caa8ff5f7fe/app/Program.cs#LL93C16-L93C16 (but better check stackoverflow)

**Reply by TheBestPessimist** [reply]
Thanks for your suggestion but now i realised that won't work either. The problem seems to be that every time i send a message to the keyboard, the "idle counter" resets to 0, so my laptop will never enter idle.

back to the drawing board :-(

**Comment by iamtimsson** [comment]
@TheBestPessimist have you tried [OpenRGB](https://github.com/fjardon/OpenRGB)?

Using the extended animation plugin, I wrote a layers animation to look like cindering kindling to emulate how this hot mess is burning itself to baser elements.

```
{
    "AllowOnlyFirst": false,
    "Brightness": 100,
    "CustomSettings": {
        "groups": [
            {
                "composer_fn": 9,
                "layers": [
                    {
                        "composer_fn": 9,
                        "effect_settings": {
                            "AllowOnlyFirst": false,
                            "Brightness": 50,
                            "CustomName": "",
                            "CustomSettings": {
                                "h": 1.0,
                                "invert_colors": false,
                                "w": 1.0,
                                "x": 0.0,
                                "y": 0.0
                            },
                            "EffectClassName": "Mask",
                            "FPS": 60,
                            "RandomColors": false,
                            "Slider2Val": 1,
                            "Speed": 1,
                            "Temperature": 0,
                            "Tint": 0,
                            "UserColors": [
                                255,
                                255
                            ]
                        }
                    },
                    {
                        "composer_fn": 6,
                        "effect_settings": {
                            "AllowOnlyFirst": false,
                            "Brightness": 100,
                            "CustomName": "",
                            "EffectClassName": "Mosaic",
                            "FPS": 60,
                            "RandomColors": false,
                            "Slider2Val": 147,
                            "Speed": 4,
                            "Temperature": 0,
                            "Tint": 0,
                            "UserColors": [
                                4351,
                                8447
                            ]
                        }
                    },
                    {
                        "composer_fn": 6,
                        "effect_settings": {
                            "AllowOnlyFirst": false,
                            "Brightness": 80,
                            "CustomName": "",
                            "CustomSettings": {
                                "lightning_mode": 1
                            },
                            "EffectClassName": "Lightning",
                            "FPS": 60,
                            "RandomColors": false,
                            "Slider2Val": 2,
                            "Speed": 69,
                            "Temperature": 0,
                            "Tint": 0,
                            "UserColors": [
                                12543
                            ]
                        }
                    }
                ]
            }
        ]
    },
    "EffectClassName": "Layers",
    "FPS": 60,
    "RandomColors": false,
    "Slider2Val": 1,
    "Speed": 1,
    "Temperature": 0,
    "Tint": 0
}
``` 

I call it burninator
it has a base layer of red, good old #ff0000 with a medium speed mosaic of redorange topped by a lightningflash of yellow, all in decreasing appearance, random across all lighting zones.
OpenRGB even set the leds on my redragon mouse came in a keyboard bundle LOL

**Reply by TheBestPessimist** [reply]
I tried OpenRGB but i could not make it work with my laptop (Asus ROG Flow X16 2023)
